### PR TITLE
Support .net6 runtime assemblies, new Build system, and Legacy Fake Codebase Separation

### PR DESCRIPTION
**Pull request by yazeedobaid**
### Introduction

This PR introduce the following modifications to Fake codebase and the build system:

1. .Net6 Runtime Assemblies Support
2. Legacy Fake Codebase Separation
3. Using GitHub Actions for Build and Test
4. Modifications to `build.fsx` File

Each point is explained in the following sections.

### .Net6 Runtime Assemblies Support

This PR adds support for .NET 6 runtime assemblies in Fake runner, reported in issue [#2588](https://github.com/fsprojects/FAKE/issues/2588). As pointed out in the issue discussion, Fake runner pins its execution to NETSTANDARD2.0, so adding a dependency that supports .NET5 or .NET6 only will fail. The move from NETSTANDARD2.0 to .NET6 is not straightforward, since in NETSTANDARD2.0 runtime assemblies that are needed by the script are installed from a NuGet package called `NETStandard.Library`. However, in .NET5 and .NET6 it is not the same.

To support new runtime assemblies, this PR make changes to `Fake.Runtime` project, the changes introduce a pinned framework in which Fake runner will use always in resolving dependencies and framework restrictions. Since .NET6 is the next LTS release, it was used instead of Netstandard2.0 which was used previously. Secondly, it will try to resolve the runtime assemblies as follows: 

1.  It first try to find a `global.json` file from script directory. If it find one it will checks the SDK version in it.
2. If SDK version found is .NET6, then it will try to locate the installed .NET6 runtime assemblies from installed SDKs.
3. Otherwise, it will default to current behavior of using NetStandard2.0 assemblies obtained from `NETStandard.Library` NuGet package.

So, to use .NET6 runtime assemblies, a `global.json` file must exists with SDK version set to .NET6. Otherwise, it will default to using Netstandard2.0 assemblies.

For the case reported in 2588 issue, [the following gist](https://www.notion.so/2336cf230b9c6137da50afd740093965) provide a simple script that demonstrate how to use .NET6 runtime assemblies by including a `global.json` file with .NET6 SDK specified.

The benefits of this solution is the switch between runtime assemblies will now be much easier. For example, when .NET6 is officially released and adopted more. We can drop .Netstandard2.0 and use by default .NET6 - for that Fake runner will require .NET6 as a requirement - and provide a switch for next release of .NET. For example when .NET7 is released, runtime assemblies for it can be activated with a `global.json` file.

The cons of this solution as pointed out in the issue is that, Fake runner now depends on an installed SDK.

In addition to runtime assemblies reference changes. A change was also introduce to the target profile argument in FSI. In FSI, the target profile for running an interactive script is different between .NET and .Netstandard. So a switch was included to handle that by resolved assemblies version in `CompileRunner.fs`.

I added two tests in integration tests project to test running Fake runner with .Netstandard2.0 and .NET6 runtime assemblies in `Fake.DotNet.RuntimeReferenceAssemblies.fs`

→ This handle issue #2588

### Legacy Fake Codebase Separation

During the implementation, I faced an issue with dependencies. I wasn't able to detect what was happening and the solution was to remove all the references in the main group in `paket.dependecies` file which as I found is for the legacy Fake build. I also disabled the build and release for legacy Fake in the Azure pipeline using the environment variable to disable legacy build. My argument for this is we need to move on with the Fake 5 implementation.

Also, while doing that, I moved the current `paket.dependencies` and `paket.lock` files to legacy directory and I marked them as ignored for reference. In addition, I moved other files/directories to legacy directory as a form of separation between Fake 5 codebase and legacy Fake codebase, which are:

1. `paket.dependencies` → `src/legacy/[ignored]paket.dependencies`
2. `paket.lock` → `src/legacy/[ignored]paket.lock`
3. `legacy-build.fsx` → `src/legacy/legacy-build.fsx`
4. `.gitlab-ci.yml` → `src/legacy/[ignored].gitlab-ci.yml`
5. `.travis.yml` → `src/legacy/[ignored].travis.yml`
6. `appveyor.yml` → `src/legacy/[ignored]appveyor.yml`
7. `Legacy-FAKE.sln` → `src/legacy/Legacy-FAKE.sln`
8. `Legacy-FAKE.sln.DotSettings` → `src/legacy/Legacy-FAKE.sln.DotSettings`
9. `Samples/` → `src/legacy/Samples/`
10. `Vagrantfile` → `src/legacy/Vagrantfile`
11. `build-web-bundles.cmd` → `src/legacy/[ignored]build-web-bundles.cmd`
12. `build-web-bundles.fsx` → `src/legacy/[ignored]build-web-bundles.fsx`

> For the CI/CD files, I moved them and consider them as legacy since this PR also introduce the usage of GitHub actions for build, test, and preparation for a release as pointed out in the section below.
> 

Also, the following Paket files have been marked as ignored to not picked up by Paket:

1. `src/legacy/FAKE/paket.references` → `src/legacy/FAKE/[ignored]paket.references`
2. `src/legacy/Fake.Experimental/ignored]paket.references`
3. `src/legacy/Fake.FluentMigrator/ignored]paket.references`
4. `src/legacy/Fake.Gallio/ignored]paket.references`
5. `src/legacy/Fake.IIS/ignored]paket.references`
6. `src/legacy/Fake.SQL/ignored]paket.references`
7. `src/legacy/FakeLib/ignored]paket.references`
8. `src/legacy/FsCheck.Fake/ignored]paket.references`
9. `src/legacy/Test.FAKECore/ignored]paket.references`
10. `src/legacy/Test.Git/ignored]paket.references`

By these changes and changes to `build.fsx` script, legacy Fake release will be discontinued.

### Using GitHub Actions for Build and Test

This PR also adds two GitHub actions for build, test, and preparing for a release. Each one is explained as follows:

- The `build_and_test` action builds the repository from PRs and Pushes to `release/next` branch on three OSs which simulate Appveyor, Travis, and Azure pipelines. It call the build script with the `Default` target which build the project and run the tests.
    
    The steps before calling the build script pack the Fake runner and replace the current `fake-cli` DotNet tool in it with produced version. I did this for two reasons:
    
    1. Dogfood the action with the latest changes in runner or repository at whole.
    2. Modifying the `Paket.Core`  version between runner versions will not enable build system to succeed. As is happening currently with the Azure pipeline. It fails due to changes in `paket.lock` file which was produced by Paket with version `6.0.13` which adds target frameworks for .Net6. And previous Fake runner uses older version of Paket which don't understand .Net6 target frameworks.
- The `prepare_release` action is intended to be run on pushes on `release/next` and/or `main` branches for preparing artifacts for a release using Windows OS which simulates current Azure CI pipeline. It uses similar steps as `build_and_test` action except it run build script with `Release_BuildAndTest` target to produce artifacts and collect them.
    
    Not sure if this actions should be run on `release/next` or `main` branches?! If we run it on `release/next` we will get artifacts published for each change. So this will enable feedback and testing more frequent. On `main` branch, it will be when a merge from `release/next` to `main` is done for releasing, which is less frequent. Currently I set the trigger to pushes on `release/next` branch.
    

These two actions can be used as in-house build and test for Fake. For release, we can still use current Azure release pipeline but give it the artifacts from `prepare_release` action.

The [following run](https://github.com/fsprojects/FAKE/actions/runs/1380662196) has the artifacts produced from `prepare_release` action - at the bottom of the page.

→ This handles issue #2607

 

### Modifications to `build.fsx` File

I have tried to organize the build script a little pit and while separating the legacy build. These include: 

1. the `#if BOOTSTRAP` directive has been removed since GitHub actions uses caching step which caches dependencies.  And while working on this PR, changes to build script dependencies didn't effect until I removed it.
2. The `.fake/build.fsx/intellisense.fsx` file and call to load it in build script has been removed since it was needed for legacy build.
3. The `publish` method now copy build artifacts to a directory to be picked up by release action.
4. Creation of Debian package command no uses `--no-restore` option since it seems .NET6 SDK is not supported yet in Debian distribution as pointed out [here](https://docs.microsoft.com/en-us/dotnet/core/install/linux-debian). Without this option, the `dotnet deb` command will try to use .NET5 references and will fail since they are not installed in build server and we use .NET6.
5. Set build version on GitHub actions to use `RunId` environment variable and commit SHA as a metadata for version. 

### Other Modifications

- The dependency `Microsoft.DotNet.PlatformAbstractions` has been removed since it is deprecated in .NET6 as pointed out [here](https://docs.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/5.0/platformabstractions-package-removed)
- A handle for native assemblies that doesn't include version specific runtime identifier was added to Fake runtime. The issue rises in integration tests that call SQLite. Specifically, the error was `Unable to load DLL 'SQLite.Interop.dll': The specified module could not be found.` .The `SQLite.Interop` library uses runtime identifier without a version specific. On windows for example, it was **win-x64**, runtime graph uses version specific runtime identifier, on my machine it was **win10-x64**. I referred to [this documentation](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog) page from Microsoft.

### ToDos

If this PR is accepted the following ToDos are still needed to be done:

- [x] Add status badges to README file for GitHub actions.
- [x] Disable trigger on Azure CI pipeline.
- [ ] Test release pipeline for picking up artifacts from GitHub actually works.
- [x] Add documentation for .NET6 runtime assembly usage.

### Summary

Sorry for making this PR very long and full of modifications, but all issues handled in this PR are related to each other in some way. I tried to document every modification I did in it. Please have a look at the proposal for supporting .NET6 runtime assemblies, new build system, and modification to separate legacy codebase from Fake 5 codebase. Your feedback is highly appreciated.

Thanks

## TODO

Feel free to open the PR and ask for help

- [ ] New (API-)documentation for new features exist (Note: API-docs are enough, additional docs are in `help/markdown`)
- [x] unit or integration test exists (or short reasoning why it doesn't make sense)
  
  > Note: Consider using the `CreateProcess` API which can be tested more easily, see https://github.com/fsharp/FAKE/pull/2131/files#diff-4fb4a77e110fbbe8210205dfe022389b for an example (the changes in the `DotNet.Testing.NUnit` module)
  
- [x] boy scout rule: "leave the code behind in a better state than you found it" (fix warnings, obsolete members or code-style in the places you worked in)     
- [ ] (if new module) the module has been linked from the "Modules" menu, edit `help/templates/template.cshtml`, linking to the API-reference is fine.
- [ ] (if new module) the module is in the correct namespace
- [ ] (if new module) the module is added to Fake.sln (`dotnet sln Fake.sln add src/app/Fake.*/Fake.*.fsproj`)
- [x] Fake 5 [API guideline](https://fake.build/contributing.html#API-Design-Guidelines) is honored

### CODE REVIEW

**Comment by dsyme** [comment]
Great to see the work in progress!

File: src/app/Fake.Runtime/FakeRuntime.fs
Comment (matthid) [inline src/app/Fake.Runtime/FakeRuntime.fs:181]: I'm not sure if this is how it is supposed to work, I'd assume some libraries might still run into issues.
Basically the runtime identifiers are organized like a tree (https://docs.microsoft.com/en-us/dotnet/core/rid-catalog), last time I checked those RIDs where pushed with a NuGet package (you should find the logic in the Paket.Core code. This might have changed by now.
But basically we should know that a win-x64 is compatible on a win10-x64 runtime and take that (disclaimer: I wrote that code in Paket and it might have changed by now or is no longer working)

Of course we can release it as-is and see if someone complains.

**Review by matthid — APPROVED** [review]
Good job, Nice!
A lot less changes than I would have expected. I guess there was nevertheless a bit of pain involved to get it running, sorry for that :)

**Comment by dsyme** [comment]
@yazeedobaid This is amazing work! Do you need further review before merging?

File: src/app/Fake.Runtime/RuntimeReferenceAssemblies.fs
Comment (matthid) [inline src/app/Fake.Runtime/RuntimeReferenceAssemblies.fs:16]: While I think the implementation is fine, I personally find that name a bit confusing.
Historically, I used the following terms: (Disclaimer: I'm not sure if that is technically the most accurate, see https://docs.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies) 
- Reference Assembly: This term is for assemblies we can only use for compilation
- Runtime Assembly: This indicates assemblies we might need to load dynamically on-demand (and potentially tell the runtime where to find them)

It is obvious that you cannot use Reference Assemblies when running the script (after compilation), because they don't actually contain any code.
It is not so obvious that you should prefer Reference Assemblies when compiling (if they are available in the NuGet package / SDK), IIRC I run into errors as well with some packages when trying to use Runtime Assemblies for compilation. Note: Most NuGet packages don't include reference assemblies, in this case using the runtime assemblies is fine and they will end up in the 'reference assemblies' list (and that is what the NuGet spec says), but all that logic is in Paket.Core. For the SDK we need to do it 'manually' unfortunately/

Given this, I find it confusing to have "RuntimeReferenceAssemblies", my suggestion would be "SdkAssemblyResolver" or something like that?

File: src/app/Fake.Runtime/RuntimeReferenceAssemblies.fs
Comment (matthid) [inline src/app/Fake.Runtime/RuntimeReferenceAssemblies.fs:27]: Should this be called "FrameworkReferenceAssemblies"?

File: src/app/Fake.Runtime/RuntimeReferenceAssemblies.fs
Comment (matthid) [inline src/app/Fake.Runtime/RuntimeReferenceAssemblies.fs:55]: "ReferenceAssemblies"?

**Comment by dsyme** [comment]
@yazeedobaid I'd love to see this merged, I think I just hit a need for net6 support.  Can we proceed? Thanks

**Comment by yazeedobaid** [comment]
@matthid Thanks for the review!
I agree with the naming changes you suggested. They are much cleaner and remove the confusion between compiled and runtime assemblies. I got drained in the changes and didn't focus on the naming side, and I picked up runtime assemblies!

So I renamed the class to `SdkAssemblyResolver` and used `SdkVersion` instead of pinned Sdk version, which indicates the LTS version of the .NET SDK.

File: src/app/Fake.Runtime/FakeRuntime.fs
Comment (yazeedobaid) [reply src/app/Fake.Runtime/FakeRuntime.fs:181]: For the native assemblies not sure why it breaks, but the two changes I think affect it are:
1. Either `Paket.Core`, in which we updated to a major version, so it may have some breaking changes in that area.
2. Or the removal of `Microsoft.DotNet.PlatformAbstractions` since it is now deprecated by Microsoft and using built-in runtime identifier property.

As you pointed out it should resolve the native package since it is a graph and `win10-x64` is a child of `win-x64`. When I debug it, the `InstallModel` resolved it with `win10-x64` but the runtime identifier from SDK was `win-x64`, and `InstallModel` didn't return anything in that configuration.

I'm not familiar with this area, if you have any suggestions I'm all open.

**Comment by yazeedobaid** [comment]
> @yazeedobaid I'd love to see this merged, I think I just hit a need for net6 support. Can we proceed? Thanks

@dsyme Yes, we can! I'm done from my side. I think we can go with an alpha release - 5.20.5-alpha.
Could you or @matthid please merge it, the merge button is not active for me, it seems the branch is protected and I don't have the sufficient permissions on it.
Thanks

**Comment by matthid** [comment]
@yazeedobaid Please check again, I assigned you the admin role.

**Comment by yazeedobaid** [comment]
Thanks @matthid

File: src/app/Fake.Runtime/FakeRuntime.fs
Comment (matthid) [reply src/app/Fake.Runtime/FakeRuntime.fs:181]: If you want to take a look, I'd say that if you debugged it to `InstallModel` it is most likely a Paket bug, so the easiest would be to add a unit-test there and try to solve it there and then ask @forki to make a release :)  
IIRC there are already tests for the runtime identifiers there, so it shouldn't be too hard to add this scenario (but that is always easy to write, I could easily be wrong on that)

**Comment by dsyme** [comment]
You guys rock. Really impressive to see you deliver this

**Pull request merged**
